src/Workspaces/Core/Portable/CodeActions/CodeAction.cs

```diff
@@ -249,6 +249,17 @@ protected virtual Task<Document> GetChangedDocumentAsync(CancellationToken cance
             return document;
         }
 
+        /// <summary>
+        /// Called by the CodeActions on the UI thread to determine if the CodeAction is still 
+        /// applicable and should be presented to the user.  CodeActions can override this if they 
+        /// need to do any final checking that must be performed on the UI thread (for example
+        /// accessing and querying the Visual Studio DTE).
+        /// </summary>
+        protected internal virtual bool IsApplicable(Workspace workspace)
```

Comment:
currently pattern seems, codefixprovider register a fix (codeaction) and then later filter some of fixes out. rather than than, how about (this is just an example, I am not sure API should be like this) CodeFixContext has something like RunOnUIThreadAsync so pattern looks like.

(engine - background thread) CodeFixProvider.RegisterCodeFixesAsync -> (user - background thread) CodeFixContext.RunOnUIThreadAsync(UIAction) -> (user - ui thread) do its stuff on UIAction and return T -> (user - background thread) using T do CodeFixContext.RegisterCodeFix -> and rest of it is same as before.